tests/test_addons/__init__.py

```diff
@@ -0,0 +1,217 @@
+        from . import scrapy_addons
+        with mock.patch.dict('sys.modules', {'scrapy.addons': scrapy_addons}):
+            self.assertEqual(self.manager.search_addon('addonmod3').FROM,
+                             'test_addons.scrapy_addons.addonmod3')
+            # Assert prefix priority 'project.addons' > 'scrapy.addons'
+            self.assertEqual(self.manager.search_addon('addonmod2').FROM,
+                             'test_addons.project.addons.addonmod2')
+            # Assert prefix priority '' > 'scrapy.addons.'
+            self.assertEqual(self.manager.search_addon('addonmod').FROM,
+                             'test_addons.addonmod')
+
+    def test_load_cfg(self):
+        manager = AddonManager()
+        manager.load_cfg(self.TESTCFGPATH)
+        six.assertCountEqual(self, manager, ['GoodAddon', 'AddonModule'])
+        six.assertCountEqual(self, manager.used_keys.values(), ['randomname',
+                              'tests.test_addons.addons.GoodAddon'])
+        self.assertIsInstance(manager['GoodAddon'], addons.GoodAddon)
+        six.assertCountEqual(self, manager.configs['GoodAddon'], ['key'])
+        self.assertEqual(manager.configs['GoodAddon']['key'], 'val1')
+        # XXX: Check module equality, see above
+        self.assertEqual(manager['AddonModule'].NAME, addonmod.NAME)
+        six.assertCountEqual(self, manager.configs['AddonModule'],
+                                   ['_source', 'key'])
+        self.assertEqual(manager.configs['AddonModule']['key'], 'val2')
+
+    def test_load_module(self):
+        manager = AddonManager()
+        manager.load_module(settings)
+        six.assertCountEqual(self, manager, ['GoodAddon', 'AddonModule'])
+        six.assertCountEqual(self, manager.used_keys.values(),
+                                   ['addonmodviasource',
+                                    'tests.test_addons.addons.GoodAddon'])
+        self.assertIsInstance(manager['GoodAddon'], addons.GoodAddon)
+        six.assertCountEqual(self, manager.configs['GoodAddon'],
+                                   ['_name', 'key'])
+        self.assertEqual(manager.configs['GoodAddon']['key'], 'val2')
+        # XXX: Check module equality, see above
+        self.assertEqual(manager['AddonModule'].NAME, addonmod.NAME)
+        six.assertCountEqual(self, manager.configs['AddonModule'],
+                                   ['_source', 'key'])
+        self.assertEqual(manager.configs['AddonModule']['key'], 'val1')
+
+    def test_callbacks(self):
+        class FirstAddon(Addon):
+            NAME = 'FirstAddon'
+            VERSION = '1.0'
+        class SecondAddon(Addon):
+            NAME = 'SecondAddon'
+            VERSION = '1.0'
+        first_addon = FirstAddon()
+        second_addon = SecondAddon()
+
+        manager = AddonManager()
+        manager.add(first_addon, {'test': 'first'})
+        manager.add(second_addon, {'test': 'second'})
+        crawler = "FakeCrawler"
```

Comment:
Using `scrapy.utils.test.get_crawler()` here threw errors I don't quite understand in the py33 test environment: https://travis-ci.org/scrapy/scrapy/builds/69558132